compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt

```diff
@@ -221,6 +217,302 @@ internal class UntilHandler(private val context: CommonBackendContext, private v
+            }
+            val stepArgValueAsLong = stepArgExpression.constLongValue
+            val checkedStepExpression = when {
+                stepArgValueAsLong == null -> {
+                    // Step argument is not a constant.
+                    val stepPositiveCheck = irCall(stepGreaterFun).apply {
+                        putValueArgument(0, stepArgExpression)
+                        putValueArgument(1, zeroStep)
+                    }
+                    irIfThenElse(
+                        stepType,
+                        stepPositiveCheck,
+                        stepArgExpression,
+                        throwIllegalStepExceptionCall
+                    )
+                }
+                stepArgValueAsLong > 0L ->
+                    // Step argument is a positive constant and is valid.
+                    stepArgExpression
+                else ->
+                    // Step argument is a non-positive constant and is invalid, directly throw the exception.
+                    throwIllegalStepExceptionCall
+            }
+
+            // While the `step` function accepts positive values, the "step" value in the progression depends on the direction of the
+            // nested progression. For example, in `10 downTo 1 step 2`, the nested progression is `10 downTo 1` which is decreasing,
+            // therefore the step used should be negated (-2).
+            //
+            // If we don't know the direction of the nested progression (e.g., `someProgression() step 2`) then we have to check its value
+            // first to determine whether to negate.
+            var nestedStepVar: IrVariable? = null
+            var checkedStepVar: IrVariable? = null
+            val checkedAndMaybeNegatedStep = when (nestedInfo.direction) {
+                ProgressionDirection.INCREASING -> checkedStepExpression
+                ProgressionDirection.DECREASING -> checkedStepExpression.negate()
+                ProgressionDirection.UNKNOWN -> {
+                    // Check value of nested step and negate step arg if needed: `if (nestedStep > 0) nestedStep else -nestedStep`
+                    // A temporary variable is created only if necessary, so we can preserve the evaluation order.
+                    val nestedStep = nestedInfo.step
+                    val nestedStepExpression = if (nestedStep.canHaveSideEffects) {
+                        nestedStepVar = scope.createTemporaryVariable(nestedStep, nameHint = "nestedStep")
+                        irGet(nestedStepVar)
+                    } else {
+                        nestedStep
+                    }
+                    val nestedStepPositiveCheck = irCall(stepGreaterFun).apply {
+                        putValueArgument(0, nestedStepExpression)
+                        putValueArgument(1, zeroStep)
+                    }
+
+                    checkedStepVar = scope.createTemporaryVariable(checkedStepExpression, nameHint = "checkedStep")
```

Comment:
The only variable that is created even if the value is constant - is this for a reason? Looks like it can be constant...

Reply:
Missed that, thanks!